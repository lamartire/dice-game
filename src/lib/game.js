import { random, round } from 'lodash'
import md5 from 'md5'

export class Game {
  constructor() {
    this.balance = 100
    this.number = null
    this.hash = null
    this.played = false
  }

  startNew() {
    this.played = false
    this.number = this.generateNumber()
    this.hash = this.generateHash(this.number)
  }

  generateNumber() {
    return random(0, 100)
  }

  takeFreeCredits() {
    if (this.balance > 0) {
      throw new Error('Free credits avilable only if balance equals to 0!')
    }

    this.balance = 100
  }

  calculateForecast(number, destination) {
    const chance = destination === 'hi' ? 100 - number : number

    return {
      chance,
      payout: round(100 / chance, 2),
    }
  }

  generateHash(number) {
    return md5(number)
  }

  validateHash(hash) {
    return this.hash === hash
  }

  makeBet(number, amount, destination) {
    if (amount > this.balance) {
      throw new Error('Amount of bet must not be more than user balance!')
    } else if (amount <= 0) {
      throw new Error('Amount of bet musy be more than 0!')
    }

    if (number >= 100) {
      throw new Error('Number must not be more than 100!')
    } else if (number <= 0) {
      throw new Error('Number must not be less than 1!')
    }

    this.played = true

    if (destination === 'lo' && this.number >= number) {
      this.processBet(number, amount, destination)

      return true
    } else if (destination === 'hi' && this.number <= number) {
      this.processBet(number, amount, destination)

      return true
    } else {
      this.withdrawBetAmount(number, amount)

      return false
    }
  }

  processBet(number, amount, destination) {
    const { payout } = this.calculateForecast(number, destination)

    this.balance = this.balance + amount * payout
  }

  withdrawBetAmount(amount) {
    this.balance = this.balance - amount
  }
}
