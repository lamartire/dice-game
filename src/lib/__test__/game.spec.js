import { Game } from '../game'

describe('pure methods', () => {
  const game = new Game()

  it('generateNumber: returns random number between 0 and 100', () => {
    const randomNumber = game.generateNumber()

    expect(randomNumber >= 0 && randomNumber <= 100).toBeTruthy()
  })

  it('generateHash: generates hash on given number', () => {
    expect(game.generateHash(1)).toEqual(expect.any(String))
  })

  it('validateHash: validate hash with current game hash', () => {
    expect()
  })

  describe('calculateForecast: returns expected chance and payout', () => {
    it('Bet 50 Hi Chance of Winning: 50% Payout: 2x ', () => {
      expect(game.calculateForecast(50, 'hi')).toMatchSnapshot()
    })

    it('Bet 50 Lo Chance of Winning: 50% Payout: 2x ', () => {
      expect(game.calculateForecast(50, 'lo')).toMatchSnapshot()
    })

    it('Bet 1 Lo Chance of Winning 1% Payout: 100x', () => {
      expect(game.calculateForecast(1, 'lo')).toMatchSnapshot()
    })

    it('Bet 99 Hi Chance of Winning 1% Payout: 100x', () => {
      expect(game.calculateForecast(99, 'hi')).toMatchSnapshot()
    })

    it('Bet 1 Hi Chance of Winning 99% Payout: 1.01x', () => {
      expect(game.calculateForecast(1, 'hi')).toMatchSnapshot()
    })

    it('Bet 100 Lo Chance of Winning 99% Payout: 1.01x', () => {
      expect(game.calculateForecast(99, 'lo')).toMatchSnapshot()
    })

    it('Bet 10 Hi Chance of Winning 90% Payout: 1.11x', () => {
      expect(game.calculateForecast(10, 'hi')).toMatchSnapshot()
    })

    it('Bet 10 Lo Chance of Winning 10% Payout: 10x', () => {
      expect(game.calculateForecast(10, 'lo')).toMatchSnapshot()
    })
  })
})

describe('methods with side-effects', () => {
  describe('takeFreeCredits', () => {
    const game = new Game()

    it('throws if player balance not equals to 0', done => {
      try {
        game.takeFreeCredits()
      } catch (err) {
        done()
      }
    })

    it('gives free 100 credits to user', () => {
      game.balance = 0
      game.takeFreeCredits()

      expect(game.balance).toEqual(100)
    })
  })

  describe('startNew', () => {
    const game = new Game()

    it('sets number and hash', () => {
      game.startNew()

      expect(game.number).not.toBeNull()
      expect(game.hash).not.toBeNull()
    })

    it('generated hash can be validated by validateHash', () => {
      expect(game.validateHash(game.hash)).toBeTruthy()
    })
  })

  it('processBet: increase user balance by amount * payout', () => {
    const game = new Game()

    game.processBet(10, 50, 'lo')

    expect(game.balance).toEqual(600)
  })

  describe('makeBet', () => {
    const game = new Game()

    it('throws if amount more than balance', done => {
      try {
        game.makeBet(1, 200, 'lo')
      } catch (err) {
        done()
      }
    })

    it('throws if amount less or equals to 0', done => {
      try {
        game.makeBet(1, 0, 'lo')
      } catch (err) {
        done()
      }
    })

    it('throws if number more or equals to 100', done => {
      try {
        game.makeBet(100, 10, 'lo')
      } catch (err) {
        done()
      }
    })

    it('throws if number less or equals to 0', done => {
      try {
        game.makeBet(0, 10, 'lo')
      } catch (err) {
        done()
      }
    })

    it('calls process bet if player won', () => {
      game.number = 50
      game.makeBet(10, 50, 'lo')

      expect(game.balance).toEqual(600)
    })
  })

  it('withdrawBetAmount: descrease player balance by bet amount', () => {
    const game = new Game()

    game.withdrawBetAmount(20)

    expect(game.balance).toEqual(80)
  })
})
