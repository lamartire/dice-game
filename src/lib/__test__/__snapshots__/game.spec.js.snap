// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`methods with side-effects createHistoryEntry: creates history entry in game history with given parameters 1`] = `
Object {
  "amount": 10,
  "hash": Any<String>,
  "number": 1,
  "result": Any<String>,
  "total": 100,
}
`;

exports[`pure methods calculateForecast: returns expected chance and payout Bet 1 Hi Chance of Winning 99% Payout: 1.01x 1`] = `
Object {
  "chance": 99,
  "payout": 1.01,
}
`;

exports[`pure methods calculateForecast: returns expected chance and payout Bet 1 Lo Chance of Winning 1% Payout: 100x 1`] = `
Object {
  "chance": 1,
  "payout": 100,
}
`;

exports[`pure methods calculateForecast: returns expected chance and payout Bet 10 Hi Chance of Winning 90% Payout: 1.11x 1`] = `
Object {
  "chance": 90,
  "payout": 1.11,
}
`;

exports[`pure methods calculateForecast: returns expected chance and payout Bet 10 Lo Chance of Winning 10% Payout: 10x 1`] = `
Object {
  "chance": 10,
  "payout": 10,
}
`;

exports[`pure methods calculateForecast: returns expected chance and payout Bet 50 Hi Chance of Winning: 50% Payout: 2x  1`] = `
Object {
  "chance": 50,
  "payout": 2,
}
`;

exports[`pure methods calculateForecast: returns expected chance and payout Bet 50 Lo Chance of Winning: 50% Payout: 2x  1`] = `
Object {
  "chance": 50,
  "payout": 2,
}
`;

exports[`pure methods calculateForecast: returns expected chance and payout Bet 99 Hi Chance of Winning 1% Payout: 100x 1`] = `
Object {
  "chance": 1,
  "payout": 100,
}
`;

exports[`pure methods calculateForecast: returns expected chance and payout Bet 100 Lo Chance of Winning 99% Payout: 1.01x 1`] = `
Object {
  "chance": 99,
  "payout": 1.01,
}
`;
